source/fe/fe_p1nc.cc

```diff
@@ -0,0 +1,418 @@
+  mpt[1](1) = (cell->vertex(1)(1) + cell->vertex(3)(1))/2.0 ;
+
+  mpt[2](0) = (cell->vertex(0)(0) + cell->vertex(1)(0))/2.0 ;
+  mpt[2](1) = (cell->vertex(0)(1) + cell->vertex(1)(1))/2.0 ;
+
+  mpt[3](0) = (cell->vertex(2)(0) + cell->vertex(3)(0))/2.0 ;
+  mpt[3](1) = (cell->vertex(2)(1) + cell->vertex(3)(1))/2.0 ;
+
+  // center point
+  Point<2> cpt ;
+  cpt(0) = (mpt[0](0) + mpt[1](0) + mpt[2](0) + mpt[3](0))/4.0 ;
+  cpt(1) = (mpt[0](1) + mpt[1](1) + mpt[2](1) + mpt[3](1))/4.0 ;
+
+  double det ;
+  det = (mpt[0](0)-mpt[1](0))*(mpt[2](1)-mpt[3](1)) - (mpt[2](0)-mpt[3](0))*(mpt[0](1)-mpt[1](1)) ;
+
+  a[0] = ((mpt[2](1)-mpt[3](1))*(0.5) -(mpt[0](1)-mpt[1](1))*(0.5))/det ;
+  a[1] = ((mpt[2](1)-mpt[3](1))*(-0.5) -(mpt[0](1)-mpt[1](1))*(0.5))/det ;
+  a[2] = ((mpt[2](1)-mpt[3](1))*(0.5) -(mpt[0](1)-mpt[1](1))*(-0.5))/det ;
+  a[3] = ((mpt[2](1)-mpt[3](1))*(-0.5) -(mpt[0](1)-mpt[1](1))*(-0.5))/det ;
+
+  b[0] = (-(mpt[2](0)-mpt[3](0))*(0.5) +(mpt[0](0)-mpt[1](0))*(0.5))/det ;
+  b[1] = (-(mpt[2](0)-mpt[3](0))*(-0.5) +(mpt[0](0)-mpt[1](0))*(0.5))/det ;
+  b[2] = (-(mpt[2](0)-mpt[3](0))*(0.5) +(mpt[0](0)-mpt[1](0))*(-0.5))/det ;
+  b[3] = (-(mpt[2](0)-mpt[3](0))*(-0.5) +(mpt[0](0)-mpt[1](0))*(-0.5))/det ;
+
+  c[0] = 0.25 - cpt(0)*a[0] - cpt(1)*b[0] ;
+  c[1] = 0.25 - cpt(0)*a[1] - cpt(1)*b[1] ;
+  c[2] = 0.25 - cpt(0)*a[2] - cpt(1)*b[2] ;
+  c[3] = 0.25 - cpt(0)*a[3] - cpt(1)*b[3] ;
+}
+
+
+
+FiniteElement<2,2>::InternalDataBase *
+FE_P1NC::get_data (const UpdateFlags update_flags,
+                   const Mapping<2,2> &,
+                   const Quadrature<2> &,
+                   dealii::internal::FEValues::FiniteElementRelatedData<2,2> &) const
+{
+  FiniteElement<2,2>::InternalDataBase *data = new FiniteElement<2,2>::InternalDataBase;
+
+  data->update_each = requires_update_flags(update_flags);
+
+  return data;
+}
+
+
+
+void
+FE_P1NC::fill_fe_values (const Triangulation<2,2>::cell_iterator           &cell,
+                         const CellSimilarity::Similarity                   ,
+                         const Quadrature<2>                               &quadrature,
+                         const Mapping<2,2>                                &mapping,
+                         const Mapping<2,2>::InternalDataBase &,
+                         const internal::FEValues::MappingRelatedData<2,2> &mapping_data,
+                         const FiniteElement<2,2>::InternalDataBase        &fe_internal,
+                         internal::FEValues::FiniteElementRelatedData<2,2> &output_data) const
+{
+  const UpdateFlags flags(fe_internal.update_each) ;
+
+  const unsigned int n_q_points = mapping_data.quadrature_points.size();
+
+  std::vector<double> values(flags & update_values ? this->dofs_per_cell : 0);
+  std::vector<Tensor<1,2> > grads(flags & update_gradients ? this->dofs_per_cell : 0);
+
+
+  // linear shape with a half value: phi(x,y) = ax + by + c
+  std::vector<double> a(4), b(4), c(4);
+  get_linear_shape (cell, a, b, c);
```

Comment:
this can be written more efficiently once you convert `get_linear_shape` to return a `std_cxx11::array`.